File: scheduler_service/internal/domain/execution.go
package domain

import (
    "time"

    "go.mongodb.org/mongo-driver/bson"
    "go.mongodb.org/mongo-driver/bson/primitive"
)

type Execution struct {
    ID                primitive.ObjectID `bson:"_id,omitempty"`
    UserId            primitive.ObjectID `bson:"userId"`
    FunctionId        primitive.ObjectID `bson:"functionId"`
    FunctionVersionId primitive.ObjectID `bson:"functionVersionId"`
    TriggerType       string             `bson:"triggerType"`
    ReplayOf          *primitive.ObjectID `bson:"replayOf,omitempty"`
    Status            string             `bson:"status"`
    InputPayload      bson.M             `bson:"inputPayload"`
    RuntimeVersion    string             `bson:"runtimeVersion"`
    ContainerImageTag string             `bson:"containerImageTag"`
    TimeoutMs         int                `bson:"timeoutMs"`
    MemoryLimitMb     int                `bson:"memoryLimitMb"`
    ExitCode          *int               `bson:"exitCode,omitempty"`
    ErrorMessage      *string            `bson:"errorMessage,omitempty"`
    StartedAt         *time.Time         `bson:"startedAt,omitempty"`
    CompletedAt       *time.Time         `bson:"completedAt,omitempty"`
    CreatedAt         time.Time          `bson:"createdAt"`
}

File: gateway_service/src/models/Execution.js
import mongoose from 'mongoose';

const executionSchema = new mongoose.Schema(
  {
    userId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: true,
      index: true,
    },
    functionId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Function',
      required: true,
      index: true,
    },
    functionVersionId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'FunctionVersion',
      required: true,
      index: true,
    },
    triggerType: {
      type: String,
      required: true,
      trim: true,
    },
    replayOf: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Execution',
      default: null,
    },
    status: {
      type: String,
      required: true,
      trim: true,
      index: true,
    },
    inputPayload: {
      type: mongoose.Schema.Types.Mixed,
      default: {},
    },
    runtimeVersion: {
      type: String,
      required: true,
      trim: true,
    },
    containerImageTag: {
      type: String,
      required: true,
      trim: true,
    },
    timeoutMs: {
      type: Number,
      required: true,
      min: 0,
    },
    memoryLimitMb: {
      type: Number,
      required: true,
      min: 0,
    },
    exitCode: {
      type: Number,
      default: null,
    },
    errorMessage: {
      type: String,
      default: null,
    },
    startedAt: {
      type: Date,
      default: null,
    },
    completedAt: {
      type: Date,
      default: null,
    },
    createdAt: {
      type: Date,
      required: true,
      default: Date.now,
      index: true,
    },
  },
  {
    versionKey: false,
  },
);

export default mongoose.model('Execution', executionSchema);
